components/server/src/authorization/spicedb.ts

```diff
@@ -40,63 +35,41 @@ export function spiceDBConfigFromEnv(): SpiceDBClientConfig | undefined {
     };
 }
 
-function spicedbClientFromConfig(config: SpiceDBClientConfig): Client {
-    const clientOptions: grpc.ClientOptions = {
-        "grpc.client_idle_timeout_ms": 10000, // this ensures a connection is not stuck in the "READY" state for too long. Required for the reconnect logic below.
-        "grpc.max_reconnect_backoff_ms": 5000, // default: 12000
-    };
-
-    return v1.NewClient(
-        config.token,
-        config.address,
-        v1.ClientSecurity.INSECURE_PLAINTEXT_CREDENTIALS,
-        undefined,
-        clientOptions,
-    ) as Client;
-}
-
-export class CachingSpiceDBClientProvider implements SpiceDBClientProvider {
+export class SpiceDBClientProvider {
     private client: Client | undefined;
 
-    private readonly interceptors: grpc.Interceptor[] = [];
-
     constructor(
-        private readonly _clientConfig: SpiceDBClientConfig,
-        private readonly clientCallMetrics?: IClientCallMetrics | undefined,
-    ) {
-        if (this.clientCallMetrics) {
-            this.interceptors.push(createClientCallMetricsInterceptor(this.clientCallMetrics));
-        }
-    }
+        private readonly clientConfig: SpiceDBClientConfig,
+        private readonly interceptors: grpc.Interceptor[] = [],
+    ) {}
 
     getClient(): SpiceDBClient {
-        let client = this.client;
-        if (!client) {
-            client = spicedbClientFromConfig(this.clientConfig);
-        } else if (client.getChannel().getConnectivityState(true) !== grpc.connectivityState.READY) {
-            // (gpl): We need this check and explicit re-connect because we observe a ~120s connection timeout without it.
-            // It's not entirely clear where that timeout comes from, but likely from the underlying transport, that is not exposed by grpc/grpc-js
-            client.close();
-
-            log.warn("[spicedb] Lost connection to SpiceDB - reconnecting...");
-
-            client = spicedbClientFromConfig(this.clientConfig);
+        if (this.client) {
+            const state = this.client.getChannel().getConnectivityState(true);
+            if (state === grpc.connectivityState.TRANSIENT_FAILURE || state === grpc.connectivityState.SHUTDOWN) {
```

Comment:
this is the most important change. We are explicitly listing the bad states on which we want to discard the connection.

Reply:
previously we were checking for RUNNING and treated all other states as bad. `CONNECTING` and `IDLE` connections were wrongly discarded here.

Reply:
@svenefftinge While TRANSIENT_FAILURE and SHUTDOWN look ok-ish, the bug I chased down last week was around connections stuck in `CONNECTING` for 2 mins, so I think we want to opportunistically re-create that connection (at least after a timeout?).
Should we sync here?

Reply:
I could not reproduce that behavior but see that connections are discarded all the time because of `CONNECTING`. If I don't discard them they just work.

Reply:
> If I don't discard them they just work.

If we always discarded too early that make sense. What's weird is why I never had that behavior while testing last week. And also, why the old behavior does not show up here, then. :grimacing: 

I will test in this preview a bit...

Reply:
I wonder why do we don't anything of it in IDE integrations 🤔 We do use @grpc/grpc-js but I was under impression that it will deal with reconnections and so on. A client only need to create generated interface and deal with failures.

Reply:
> I wonder why do we don't anything of it in IDE integrations

It totally depends on the Client/Server combination:
- we own most of the gRPC services we consume ourselves, so we can properly configure keepalive and server-side timeouts etc. to our needs
- a lot of our client use `nice-grpc` nowadays, and those seem to work fine for some reason (worth investigating?)
- for spicedb:
  - we don't control the server side
  - we have limited control over the client, as we consume the hand-written (?) client published by authzed